Chartboost/ChartboostBannerCustomEvent.m

```diff
@@ -6,58 +6,38 @@
-    NSString *appSignature = [info objectForKey:@"appSignature"];
-    MPLogAdEvent([MPLogEvent adLoadAttemptForAdapter:NSStringFromClass(self.class) dspCreativeId:nil dspName:nil], self.appID);
-    
-    if ([self.appID length] == 0 || [appSignature length] == 0) {
-        NSError *error = [NSError errorWithCode:MOPUBErrorAdapterInvalid localizedDescription:@"Failed to load Chartboost banner: missing either appId or appSignature. Make sure you have a valid appId or appSignature entered on the MoPub dashboard."];
-        MPLogAdEvent([MPLogEvent adLoadFailedForAdapter:NSStringFromClass(self.class) error:error], self.appID);
-        [self.delegate bannerCustomEvent:self didFailToLoadAdWithError:error];
-        
-        return;
-    }
-    
-    if (self.banner) {
-        NSError *error = [NSError errorWithCode:MOPUBErrorAdapterInvalid localizedDescription:@"Chartboost adapter failed to load ad: requestAdWithSize called twice on the same event."];
-        MPLogAdEvent([MPLogEvent adLoadFailedForAdapter:NSStringFromClass(self.class) error:error], self.appID);
-        [self.delegate bannerCustomEvent:self didFailToLoadAdWithError:error];
-        
-        return;
-    }
-    
-    [ChartboostAdapterConfiguration updateInitializationParameters:info];
-    
     NSString *location = [info objectForKey:@"location"];
-    location = [location length] != 0 ? location: CBLocationDefault;
+    location = location.length > 0 ? location : CBLocationDefault;
     CGSize integerSize = CGSizeMake(floor(size.width), floor(size.height));
+
+    MPLogAdEvent([MPLogEvent adLoadAttemptForAdapter:NSStringFromClass(self.class) dspCreativeId:nil dspName:nil], location);
+    if (self.banner) {
+        MPLogAdEvent([MPLogEvent error:[NSError adRequestCalledTwiceOnSameEvent] message:nil], location);
+    }
     
     __weak typeof(self) weakSelf = self;
-    [[ChartboostRouter sharedRouter] startWithAppId:self.appID appSignature:appSignature completion:^(BOOL initialized) {
+    [ChartboostRouter startWithParameters:info completion:^(BOOL initialized) {
+        if (!initialized) {
+            NSError *error = [NSError adRequestFailedDueToSDKStartWithAdOfType:@"banner"];
+            MPLogAdEvent([MPLogEvent adLoadFailedForAdapter:NSStringFromClass(self.class) error:error], location);
+            [self.delegate bannerCustomEvent:self didFailToLoadAdWithError:error];
+            return;
+        }
         dispatch_async(dispatch_get_main_queue(), ^{
-            if (!weakSelf.banner) {
-                weakSelf.banner = [[CHBBanner alloc] initWithSize:integerSize location:location delegate:weakSelf];
-                weakSelf.banner.automaticallyRefreshesContent = NO;
-            }
+            weakSelf.banner.delegate = nil;
+            weakSelf.banner = [[CHBBanner alloc] initWithSize:integerSize location:location mediation:[ChartboostRouter mediation] delegate:weakSelf];
+            weakSelf.banner.automaticallyRefreshesContent = NO;
```

Comment:
Yes, but on mopub banners.
MoPub custom events are unique for each ad request. Every time mopub wants to load a new banner ad (on a new banner or to refresh an existing one) it will create a new event and we pass a new banner object there.
They handle the refresh themselves, I believe they have some settings in their dashboard.